core/query/sql.ml

```diff
@@ -0,0 +1,467 @@
+              (* Inline the query *)
+              let tables = List.map (fun (t, x) -> db#quote_field t ^ " as " ^ (string_of_table_var x)) tables in
+              let q = "(" ^ sq q ^ ") as " ^ string_of_table_var z in
+              string_of_select fields (q::tables) condition os
+          | _ -> assert false
+
+and string_of_base db one_table b =
+  let sb = string_of_base db one_table in
+    match b with
+      | `Case (c, t, e) ->
+          "case when " ^ sb c ^ " then " ^sb t ^ " else "^ sb e ^ " end"
+      | `Constant c -> Constant.string_of_constant c
+      | `Project (_var, _label as p) -> string_of_projection db one_table p
+      | `Apply (op, [l; r]) when Arithmetic.is op
+          -> Arithmetic.gen (sb l, op, sb r)
+      | `Apply (("intToString" | "stringToInt" | "intToFloat" | "floatToString"
+                | "stringToFloat"), [v]) -> sb v
+      | `Apply ("floatToInt", [v]) -> "floor("^sb v^")"
+
+      (* optimisation *)
+      | `Apply ("not", [`Empty q]) -> "exists (" ^ string_of_query db true q ^ ")"
+
+      | `Apply ("not", [v]) -> "not (" ^ sb v ^ ")"
+      | `Apply (("negate" | "negatef"), [v]) -> "-(" ^ sb v ^ ")"
+      | `Apply ("&&", [v; w]) -> "(" ^ sb v ^ ")" ^ " and " ^ "(" ^ sb w ^ ")"
+      | `Apply ("||", [v; w]) -> "(" ^ sb v ^ ")" ^ " or " ^ "(" ^ sb w ^ ")"
+      | `Apply ("==", [v; w]) -> "(" ^ sb v ^ ")" ^ " = " ^ "(" ^ sb w ^ ")"
+      | `Apply ("<>", [v; w]) -> "(" ^ sb v ^ ")" ^ " <> " ^ "(" ^ sb w ^ ")"
+      | `Apply ("<", [v; w]) -> "(" ^ sb v ^ ")" ^ " < " ^ "(" ^ sb w ^ ")"
+      | `Apply (">", [v; w]) -> "(" ^ sb v ^ ")" ^ " > " ^ "(" ^ sb w ^ ")"
+      | `Apply ("<=", [v; w]) -> "(" ^ sb v ^ ")" ^ " <= " ^ "(" ^ sb w ^ ")"
+      | `Apply (">=", [v; w]) -> "(" ^ sb v ^ ")" ^ " >= " ^ "(" ^ sb w ^ ")"
+      | `Apply ("RLIKE", [v; w]) -> "(" ^ sb v ^ ")" ^ " RLIKE " ^ "(" ^ sb w ^ ")"
+      | `Apply ("LIKE", [v; w]) -> "(" ^ sb v ^ ")" ^ " LIKE " ^ "(" ^ sb w ^ ")"
+      | `Apply (f, args) when SqlFuns.is f -> SqlFuns.name f ^ "(" ^ String.concat "," (List.map sb args) ^ ")"
+      | `Apply (f, args) -> f ^ "(" ^ String.concat "," (List.map sb args) ^ ")"
+      | `Empty q -> "not exists (" ^ string_of_query db true q ^ ")"
+      | `Length q -> "select count(*) from (" ^ string_of_query db true q ^ ") as " ^ fresh_dummy_var ()
+      | `RowNumber [] -> "1"
+      | `RowNumber ps ->
+        "row_number() over (order by " ^ String.concat "," (List.map (string_of_projection db one_table) ps) ^ ")"
+and string_of_projection db one_table (var, label) =
+  if one_table then
+    db#quote_field label
+  else
+    string_of_table_var var ^ "." ^ (db#quote_field label)
+
+let string_of_query db range q =
+  let range =
+    match range with
+      | None -> ""
+      | Some (limit, offset) -> " limit " ^string_of_int limit^" offset "^string_of_int offset
+  in
+    string_of_query db false q ^ range
+
+let prepare_clauses : Q.t -> Q.t list =
```

Comment:
THe code from here to about line 436 seems to be more about "converting Q.t to Sql.query" (with/without shredding-specific complications such as indexes) and not about SQL per se.  I wonder if this code should live in query.ml instead of sql.ml.  (I'd also like to get rid of any dependency of sql.ml on query.ml - if anything the dependency should be in the other direction, or the translation from query.t to sql.t should happen in a later module that imports both.)

Reply:
Ah yes, that makes sense. I'll look into it.

Reply:
It looks like `update` and `delete` will also need to be moved to `Query`, as they both depend on `Query.base`.
This is hardly surprising, as we already had `ordered_query` in `EvalQuery` and `unordered_query_package` in `EvalNestedQuery`

Reply:
That seems like an improvement.  The (compile) insert/delete/update functions are really about Links's interface to SQL updates, we don't have any particular syntax for SQL updates themselves at the moment.  These might eventually deserve to be in a separate (Update) module at some point but I can live with this for now.